File: 30.substring-with-concatenation-of-all-words.py
# @before-stub-for-debug-begin
from python3problem30 import *
from typing import *
# @before-stub-for-debug-end

#
# @lc app=leetcode id=30 lang=python3
#
# [30] Substring with Concatenation of All Words
#

# @lc code=start
from typing import List


class Solution:
    def findSubstring(self, s: str, words: List[str]) -> List[int]:
        # we have some string "s" that contains some number of substrings consisting of exact concatinations of list "words"
        # find all starting indicies of these substrings
        # so we want to walk through the string s and check if thats the start of one of the substrings.
        # since all words are the same length we can chop out the expected word size and check if the word matches.
        # if we have a match start going through the rest of the substring making sure we get one instance of each
        # if we do then record the start index to some list and go back to the start position and continue looking
        # char by char
        # we can also end when there isnt enough room to complete another substring len(words)*length of word

        indicies: List[int] = []
        word_length: int = len(words[0])
        substring_length: int = len(words) * word_length
        max_index = len(s) + 1 - substring_length

        for i in range(max_index):
            substring: str = s[i:i+substring_length]
            if self.checkSubString(substring, words.copy()):
                indicies.append(i)
        return indicies
    def checkSubString(self, substring: str, words:List[str]) -> bool:
        word_length: int = len(words[0])
        #print(f"substring: {substring}")
        for i in range(len(words)):
            word = substring[i*word_length:(i+1)*word_length]
            if word not in words:
                #print(f"word: {word} does not exist")
                return False
            #print(f"word: {word} exists")
            words.remove(word)
        return True

        
# @lc code=end

File: rust/src/bin/30.substring-with-concatenation-of-all-words.rs
/*
 * @lc app=leetcode id=30 lang=rust
 *
 * [30] Substring with Concatenation of All Words
 */

use std::str::Chars;

// @lc code=start
impl Solution {
    pub fn find_substring(s: String, words: Vec<String>) -> Vec<i32> {

        let mut indicies: Vec<i32> = [].to_vec();
        let word_length: usize = words[0].chars().count();
        let substring_length: i32 = (words.len() * word_length) as i32;
        let max_index: i32 = (s.chars().count() + 1) as i32 - substring_length;

        let mut valid_substrings: Vec<&str> = vec![];

        for i in 0..max_index
        {
            let substring: &str = &(s[i as usize..(i+substring_length) as usize]);
            if valid_substrings.contains(&substring){
                indicies.push(i);
            }
            else if Self::check_substring(&substring, &mut words.clone()){
                valid_substrings.push(substring);
                indicies.push(i);
            }
        }

        return indicies;
    }

    pub fn check_substring(substring: &str, words: &mut Vec<String>) -> bool {
        let word_length: usize = words[0].chars().count();
        //println!("substring: {substring}");
        for i in 0..words.len() {

            let word: &str = &(substring[i*word_length..(i+1)*word_length]);
            
            let mut found: bool = false;
            for i in 0..words.len(){
                if words[i] == word {
                    //println!("word: {word} found");
                    words.remove(i);
                    found = true;
                    break;
                }
            }

            if !found {
                //println!("word: {word} not found");
                return false;
            }

        }
        return true;
    }
}
// @lc code=end

